src/buffer.cc

```diff
@@ -204,6 +205,85 @@ Result Buffer::CompareRMSE(Buffer* buffer, float tolerance) const {
   return {};
 }
 
+std::vector<int64_t> Buffer::GetHistogramForChannel(uint32_t channel,
+                                                    uint32_t num_bins) const {
+  assert(num_bins == 256);
+  std::vector<int64_t> bins(num_bins, 0.0);
+  auto* buf_ptr = GetValues<uint8_t>();
+  auto num_channels = format_->InputNeededPerElement();
+  uint32_t channel_id = 0;
+
+  for (size_t i = 0; i < ElementCount(); ++i) {
+    for (const auto& seg : format_->GetSegments()) {
+      if (seg.IsPadding()) {
+        buf_ptr += seg.PaddingBytes();
+        continue;
+      }
+      if (channel_id == channel) {
+        assert(type::Type::IsUint8(seg.GetFormatMode(), seg.GetNumBits()));
+        const auto bin = *reinterpret_cast<const uint8_t*>(buf_ptr);
+        bins[bin]++;
+      }
+      buf_ptr += seg.SizeInBytes();
+      channel_id = (channel_id + 1) % num_channels;
+    }
+  }
+
+  return bins;
+}
+
+Result Buffer::CompareHistogramEMD(Buffer* buffer, float tolerance) const {
+  auto result = CheckCompability(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  const int num_bins = 256;
+  auto num_channels = format_->InputNeededPerElement();
+  for (auto segment : format_->GetSegments()) {
+    if (!type::Type::IsUint8(segment.GetFormatMode(), segment.GetNumBits()) ||
+        num_channels != 4)
+      return Result(
+          "EMD comparison only supports 8bit unorm format with four channels.");
+  }
+
+  std::vector<std::vector<int64_t>> histogram1, histogram2;
```

Comment:
You have simplified the code nicely below. However, I would rather normalize the histograms first according to the number of elements in *each* buffer, before you do the EMD part. In theory, the images could have difference sizes. Not sure if this can happen (nor whether we even want it to happen), but the algorithm is more future-proof if we normalize the histograms first. Thus, use vector<double> here.